src/dscanner/analysis/cyclomatic_complexity.d

```diff
@@ -50,219 +43,348 @@ final class CyclomaticComplexityCheck : BaseAnalyzer
+			return;
+
+		complexityStack[$ - 1]++;
+		funcExp.fd.accept(this);
+	}
 
-	void testComplexity(T)(T annotatable)
+	mixin VisitComplex!(AST.IfStatement);
+	mixin VisitComplex!(AST.LogicalExp);
+	mixin VisitComplex!(AST.CondExp);
+	mixin VisitComplex!(AST.CaseStatement);
+	mixin VisitComplex!(AST.CaseRangeStatement);
+	mixin VisitComplex!(AST.ReturnStatement);
+	mixin VisitComplex!(AST.ContinueStatement);
+	mixin VisitComplex!(AST.GotoStatement);
+	mixin VisitComplex!(AST.TryFinallyStatement);
+	mixin VisitComplex!(AST.ThrowExp);
+
+	private template VisitComplex(NodeType, int increase = 1)
 	{
-		auto complexity = complexityStack[$ - 1];
-		if (complexity > maxCyclomaticComplexity)
+		override void visit(NodeType nodeType)
 		{
-			addErrorMessage(annotatable, KEY, format!MESSAGE(complexity));
+			complexityStack[$ - 1] += increase;
+			super.visit(nodeType);
 		}
 	}
 
-	template VisitComplex(NodeType, int increase = 1)
+	override void visit(AST.SwitchStatement switchStatement)
+	{
+		inLoop.assumeSafeAppend ~= false;
+		scope (exit)
+		inLoop.length--;
+
+		switchStatement.condition.accept(this);
+		switchStatement._body.accept(this);
+	}
+
+	override void visit(AST.BreakStatement breakStatement)
 	{
-		override void visit(const NodeType n)
+		if (inLoop[$ - 1])
+			complexityStack[$ - 1]++;
+	}
+
+	override void visit(AST.TryCatchStatement tryCatchStatement)
+	{
+		tryCatchStatement._body.accept(this);
+
+		if (tryCatchStatement.catches !is null)
 		{
-			complexityStack[$ - 1] += increase;
-			n.accept(this);
+			foreach (catchStatement; *(tryCatchStatement.catches))
+			{
+				complexityStack[$ - 1]++;
+				catchStatement.handler.accept(this);
+			}
 		}
 	}
 
-	template VisitLoop(NodeType, int increase = 1)
+	override void visit(AST.StaticForeachStatement staticForeachStatement)
 	{
-		override void visit(const NodeType n)
+		// StaticForeachStatement visit has to be overridden in order to avoid visiting
+		// its forEachStatement member, which would increase the complexity.
+		return;
+	}
+
+	mixin VisitLoop!(AST.DoStatement);
+	mixin VisitLoop!(AST.WhileStatement);
+	mixin VisitLoop!(AST.ForStatement);
+	mixin VisitLoop!(AST.ForeachRangeStatement);
+	mixin VisitLoop!(AST.ForeachStatement);
+
+	private template VisitLoop(NodeType, int increase = 1)
+	{
+		override void visit(NodeType nodeType)
 		{
 			inLoop.assumeSafeAppend ~= true;
 			scope (exit)
-				inLoop.length--;
+			inLoop.length--;
+
 			complexityStack[$ - 1] += increase;
-			n.accept(this);
+			super.visit(nodeType);
 		}
 	}
 }
 
 unittest
 {
 	import dscanner.analysis.config : StaticAnalysisConfig, Check, disabledConfig;
+	import dscanner.analysis.helpers : assertAnalyzerWarningsDMD;
 	import std.stdio : stderr;
 
 	StaticAnalysisConfig sac = disabledConfig();
 	sac.cyclomatic_complexity = Check.enabled;
 	sac.max_cyclomatic_complexity = 0;
-	assertAnalyzerWarnings(q{
-unittest /+
-^^^^^^^^ [warn]: Cyclomatic complexity of this function is 1. +/
-{
-}
```

Comment:
Why is this test deleted?

Reply:
This test is similar to the one below - unit test only increases the complexity by 1